Adding pymongo functional tests

hectorhdzg opened the pull request:
Using similar approach as OpenTelemetry JS to run docker with MongodDB container and run tests against it, currently it will require contributors of this repo to have docker installed and running when executing the tests

Fixes #284

Oberon00 commented on ext/opentelemetry-ext-pymongo/tests/test_pymongo_functional.py:
```diff
@@ -0,0 +1,124 @@
+                check=False,
+            )
+            if process.returncode != 0:
+                logging.warning("Failed to start MongoDB container")
```
Why not just use `check=True`?

hectorhdzg replied:
Container could be already running because it was started by the user, the test didn't execute clean up while debugging, etc. I didn't want to fail the test case here, it would fail when creating MongoClient if something goes wrong anyways

hectorhdzg commented:
@toumorokoshi , @c24t  I updated the code to not do any docker related operations in the test classes, created new tox environment just for these to be able to disable easily in case there are any issue and added a dependency on docker-compose to easily add new containers if needed, let me know if you have any feedback

ocelotl commented on ext/opentelemetry-ext-docker-tests/tests/pymongo/test_pymongo_functional.py:
```diff
@@ -0,0 +1,105 @@
+            self._collection.insert_one(
+                {"name": "testName", "value": "testValue"}
+            )
+        self.validate_spans(self._span_exporter.get_finished_spans())
```
`self.validate_spans` always has `self._span_exporter.get_finished_spans()` as an argument. This probably means that `spans = self._span_exporter.get_finished_spans()` should be at the very beginning of `TestFunctionalPymongo.validate_spans` instead.

ocelotl commented on ext/opentelemetry-ext-docker-tests/tests/pymongo/test_pymongo_functional.py:
```diff
@@ -0,0 +1,105 @@
+            self.assertIsInstance(span.start_time, int)
+            self.assertIsInstance(span.end_time, int)
+        self.assertIsNotNone(pymongo_span.parent)
+        self.assertEqual(pymongo_span.parent.name, root_span.name)
```
By this point you require `pymongo_span` and `root_span` to be defined. They may not be if `spans` only includes 2 spans whose `name` attribute is `"rootSpan"` or if it includes 2 spans whose name is not. If any of those happen, you'll get a `NameError` in a test case when you want to get only `AssertionError`s. Instead for looping through `spans` looking for spans whose `name` attribute matches, it should be asserted that `spans` actually contains one span whose `name` attribute is `"rootSpan"` and another one whose `name` attribute is not.

hectorhdzg replied:
The actual test I wanted to include is that rootSpan is parent of other spans created, that is why I'm saving the reference for it so I can assert later, the check of an extra span being created in pymongo is already taken care in self.assertEqual(len(spans), 2) because all tests start with a rootSpan only, are you suggesting to add more check regarding these spans?

ocelotl replied:
Yes, what I am pointing out is that your test case requires `self._span_exporter.get_finished_spans()` to deliver you 2 spans, one named `"rootSpan"` and other not, and since it requires that, it is probably something that you want to specifically test.

c24t commented:
Sorry to hold you up @hectorhdzg. This PR LGTM after splitting off the docker tests, and it looks like you addressed @ocelotl's comment in 176cb73.

c24t commented:
@ocelotl if this LGTY we can merge it.

ocelotl commented:
Sure, I think [this](https://github.com/open-telemetry/opentelemetry-python/pull/340#discussion_r373209750) could be addressed but it is not critical. :+1:
